Scala compile error: invalid prefix TypeBound

The type is compiling fine and I'm using dotty version "3.0.0-RC2-bin-20210325-eeb8340-NIGHTLY"
If I remove the any Nop continuation, even just one of them, it will compile fine
Apologies for this long code, I am new to Dotty and I have tried my best to minimise it

## Compiler version
Scala compiler version 2.12.3 / Dotty compiler version 3.0.0-M1

## Minimized code

```scala
case class Test1(); case class Test2(); case class Test3();
case class Test4(); case class Test5(); case class Test6();

sealed abstract class DSL {
  def cont [P1 >: this.type <: DSL, P2 <: DSL](continuation: => P2) =
    Continue[P1, P2](() => this, () => continuation)
}
case class Continue [P1 <: DSL, P2 <: DSL](p1: () => P1, p2: () => P2) extends DSL

trait More[-A] {}
case class Out[C <: More[A], A](c: C, v: A) extends DSL
case class Nop() extends DSL

val decision1:Boolean = true;
val decision2:Boolean = false;

type P[
ChanA <: More[Test1|Test2],
ChanB <: More[Test3|Test4],
ChanC <: More[Test5|Test6]] =
 ((Out[ChanA,Test1] Continue ((Out[ChanB,Test3] Continue Nop)|(Out[ChanB,Test4] Continue Nop))) //works if remove first 'Continue Nop'
 | (Out[ChanA,Test2] Continue ((Out[ChanC,Test5] Continue Nop)|(Out[ChanC,Test6] Continue Nop))))


def p( chanA: More[Test1|Test2], chanB: More[Test3|Test4], chanC: More[Test5|Test6])
 :P[chanA.type,chanB.type,chanC.type] ={
      if(decision1){
         Out(chanA,Test1()) cont {
            if(decision2){
               Out(chanB,Test3()) cont Nop() //works if replace with 'Out(chanB,Test3())'
            }
            else{
               Out(chanB,Test4()) cont Nop()
            }
         }
      }
      else{
         Out(chanA,Test2()) cont {
            if(decision2){
               Out(chanC,Test5()) cont Nop()
            }
            else{
               Out(chanC,Test6()) cont Nop()
            }
         }
      }
   }
```

## Output

```scala
[error] java.lang.AssertionError: assertion failed: invalid prefix TypeBounds(OrType(AndType(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Out),List(TypeParamRef(ChanB), TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Test3))),AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Out),List(TypeParamRef(ChanB), TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Test4)))),AndType(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Out),List(TypeParamRef(ChanB), TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Test3))),AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Out),List(TypeParamRef(ChanC), TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Test6))))),OrType(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Out),List(TypeParamRef(ChanB), TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Test3))),AndType(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Out),List(TypeParamRef(ChanB), TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Test4))),AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Out),List(TypeParamRef(ChanC), TypeRef(ThisType(TypeRef(NoPrefix,module class bla)),class Test6))))))
[error] 	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] 	at dotty.tools.dotc.core.Types$NamedType.<init>(Types.scala:2023)
[error] 	at dotty.tools.dotc.core.Types$TypeRef.<init>(Types.scala:2593)
[error] 	at dotty.tools.dotc.core.Types$CachedTypeRef.<init>(Types.scala:2649)
[error] 	at dotty.tools.dotc.core.Uniques$NamedTypeUniques.newType$1(Uniques.scala:43)
[error] 	at dotty.tools.dotc.core.Uniques$NamedTypeUniques.enterIfNew(Uniques.scala:53)
[error] 	at dotty.tools.dotc.core.Types$TypeRef$.apply(Types.scala:2708)
[error] 	at dotty.tools.dotc.core.TypeComparer.compareCaptured$1(TypeComparer.scala:1437)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubArg$1(TypeComparer.scala:1468)
[error] 	at dotty.tools.dotc.core.TypeComparer.recurArgs$1(TypeComparer.scala:1475)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubArgs(TypeComparer.scala:1478)
[error] 	at dotty.tools.dotc.core.TypeComparer.loop$3(TypeComparer.scala:1071)
[error] 	at dotty.tools.dotc.core.TypeComparer.isMatchingApply$1(TypeComparer.scala:1086)
[error] 	at dotty.tools.dotc.core.TypeComparer.compareAppliedType2$1(TypeComparer.scala:1145)
[error] 	at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:556)
[error] 	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:487)
[error] 	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:369)
[error] 	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1261)
[error] 	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:480)
[error] 	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:369)
[error] 	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1261)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:180)
[error] 	at dotty.tools.dotc.core.TypeComparer.tryBaseType$1(TypeComparer.scala:742)
[error] 	at dotty.tools.dotc.core.TypeComparer.compareAppliedType2$1(TypeComparer.scala:1153)
[error] 	at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:556)
[error] 	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:487)
[error] 	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:369)
[error] 	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1261)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:180)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:190)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSameType(TypeComparer.scala:1881)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSame(TypeComparer.scala:1883)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.op$proxy2$1(ConstraintHandling.scala:222)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.isSameTypeWhenFrozen(ConstraintHandling.scala:222)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.isSameTypeWhenFrozen$(ConstraintHandling.scala:26)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSameTypeWhenFrozen(TypeComparer.scala:30)
[error] 	at dotty.tools.dotc.core.TypeComparer.singletonInterval(TypeComparer.scala:1917)
[error] 	at dotty.tools.dotc.core.TypeComparer.glbArgs(TypeComparer.scala:2053)
[error] 	at dotty.tools.dotc.core.TypeComparer.distributeAnd(TypeComparer.scala:2218)
[error] 	at dotty.tools.dotc.core.TypeComparer.andTypeGen(TypeComparer.scala:2126)
[error] 	at dotty.tools.dotc.core.TypeComparer.andType(TypeComparer.scala:2154)
[error] 	at dotty.tools.dotc.core.TypeComparer.glb(TypeComparer.scala:1962)
[error] 	at dotty.tools.dotc.core.TypeComparer$.glb(TypeComparer.scala:2637)
[error] 	at dotty.tools.dotc.core.Types$Type.$amp(Types.scala:1122)
[error] 	at dotty.tools.dotc.core.TypeOps$.mergeRefinedOrApplied$1(TypeOps.scala:244)
[error] 	at dotty.tools.dotc.core.TypeOps$.baseTp$2$$anonfun$2(TypeOps.scala:332)
[error] 	at dotty.tools.dotc.core.Types$Type.mapReduceOr(Types.scala:464)
[error] 	at dotty.tools.dotc.core.TypeOps$.baseTp$3(TypeOps.scala:332)
[error] 	at dotty.tools.dotc.core.TypeOps$.approximateOr$2$$anonfun$1(TypeOps.scala:333)
[error] 	at scala.collection.immutable.List.map(List.scala:246)
[error] 	at dotty.tools.dotc.core.TypeOps$.approximateOr$1(TypeOps.scala:333)
[error] 	at dotty.tools.dotc.core.TypeOps$.orDominator(TypeOps.scala:338)
[error] 	at dotty.tools.dotc.core.Types$OrType.join(Types.scala:3151)
[error] 	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:480)
[error] 	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:343)
[error] 	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1261)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:180)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:190)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSub(TypeComparer.scala:192)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.op$proxy1$1(ConstraintHandling.scala:221)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.isSubTypeWhenFrozen(ConstraintHandling.scala:221)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.isSubTypeWhenFrozen$(ConstraintHandling.scala:26)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubTypeWhenFrozen(TypeComparer.scala:30)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.isSubType(ConstraintHandling.scala:205)
[error] 	at dotty.tools.dotc.core.ConstraintHandling.isSubType$(ConstraintHandling.scala:26)
[error] 	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:30)
[error] 	at dotty.tools.dotc.core.TypeComparer.mergeIfSuper(TypeComparer.scala:2108)
[error] 	at dotty.tools.dotc.core.TypeComparer.mergedLub$1(TypeComparer.scala:1994)
[error] 	at dotty.tools.dotc.core.TypeComparer.lub(TypeComparer.scala:2006)
[error] 	at dotty.tools.dotc.core.TypeComparer$.lub(TypeComparer.scala:2627)
[error] 	at dotty.tools.dotc.core.Types$Type.$bar(Types.scala:1156)
[error] 	at dotty.tools.dotc.core.TypeComparer.glb(TypeComparer.scala:1934)
[error] 	at dotty.tools.dotc.core.TypeComparer.lubArgs(TypeComparer.scala:2027)
[error] 	at dotty.tools.dotc.core.TypeComparer.lubArgs(TypeComparer.scala:2029)
[error] 	at dotty.tools.dotc.core.TypeComparer$.lubArgs(TypeComparer.scala:2634)
[error] 	at dotty.tools.dotc.core.TypeOps$.mergeRefinedOrApplied$1(TypeOps.scala:234)
[error] 	at dotty.tools.dotc.core.TypeOps$.baseTp$2$$anonfun$2(TypeOps.scala:332)
[error] 	at dotty.tools.dotc.core.Types$Type.mapReduceOr(Types.scala:464)
[error] 	at dotty.tools.dotc.core.TypeOps$.baseTp$3(TypeOps.scala:332)
[error] 	at dotty.tools.dotc.core.TypeOps$.approximateOr$2$$anonfun$1(TypeOps.scala:333)
[error] 	at scala.collection.immutable.List.map(List.scala:246)
[error] 	at dotty.tools.dotc.core.TypeOps$.approximateOr$1(TypeOps.scala:333)
[error] 	at dotty.tools.dotc.core.TypeOps$.orDominator(TypeOps.scala:338)
[error] 	at dotty.tools.dotc.core.Types$OrType.join(Types.scala:3151)
[error] 	at dotty.tools.dotc.core.Types$Type.classSymbol(Types.scala:509)
[error] 	at dotty.tools.dotc.core.TypeErasure.dotty$tools$dotc$core$TypeErasure$$apply(TypeErasure.scala:594)
[error] 	at dotty.tools.dotc.core.TypeErasure.eraseResult(TypeErasure.scala:764)
[error] 	at dotty.tools.dotc.core.TypeErasure.dotty$tools$dotc$core$TypeErasure$$apply(TypeErasure.scala:634)
[error] 	at dotty.tools.dotc.core.TypeErasure.eraseInfo(TypeErasure.scala:706)
[error] 	at dotty.tools.dotc.core.TypeErasure$.transformInfo(TypeErasure.scala:243)
[error] 	at dotty.tools.dotc.transform.Erasure.transform(Erasure.scala:85)
[error] 	at dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:832)
[error] 	at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:881)
[error] 	at dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:122)
[error] 	at dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:116)
[error] 	at dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:109)
[error] 	at dotty.tools.dotc.core.Symbols$.toDenot(Symbols.scala:494)
[error] 	at dotty.tools.dotc.transform.Erasure$Typer.dropConstructorProxies$$anonfun$1(Erasure.scala:1052)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.immutable.List.foreach(List.scala:333)
[error] 	at dotty.tools.dotc.core.Scopes$Scope.foreach(Scopes.scala:94)
[error] 	at dotty.tools.dotc.transform.Erasure$Typer.dropConstructorProxies(Erasure.scala:1052)
[error] 	at dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1058)
[error] 	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2661)
[error] 	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2665)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2734)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2804)
[error] 	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2826)
[error] 	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2876)
[error] 	at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1068)
[error] 	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2455)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2706)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2735)
[error] 	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2804)
[error] 	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2920)
[error] 	at dotty.tools.dotc.transform.Erasure.run(Erasure.scala:132)
[error] 	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
[error] 	at scala.collection.immutable.List.map(List.scala:246)
[error] 	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
[error] 	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] 	at dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
[error] 	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] 	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] 	at dotty.tools.dotc.Run.compileUnits(Run.scala:230)
[error] 	at dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error] 	at dotty.tools.dotc.Run.compile(Run.scala:150)
[error] 	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] 	at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] 	at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] 	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:92)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
[error] 	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:573)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:573)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
[error] 	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] 	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] 	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] 	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] 	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] 	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
[error] 	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] 	at sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:573)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] 	at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2200)
[error] 	at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2157)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:40)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] 	at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] 	at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2153)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
[error] 	at java.base/java.lang.Thread.run(Thread.java:832)
```

## Expectation

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -1438,15 +1438,16 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
          */
         def compareCaptured(arg1: TypeBounds, arg2: Type) = tparam match {
           case tparam: Symbol =>
-            if (leftRoot.isStable || (ctx.isAfterTyper || ctx.mode.is(Mode.TypevarsMissContext))
-                && leftRoot.member(tparam.name).exists) {
+            if (leftRoot.isStable || ctx.isAfterTyper || ctx.mode.is(Mode.TypevarsMissContext))
+                && leftRoot.isValueType
+                && leftRoot.member(tparam.name).exists
+            then
               val captured = TypeRef(leftRoot, tparam)
               try isSubArg(captured, arg2)
               catch case ex: TypeError =>
                 // The captured reference could be illegal and cause a
                 // TypeError to be thrown in argDenot
                 false
-            }
             else if (v > 0)
               isSubType(paramBounds(tparam).hi, arg2)
             else if (v < 0)

--- a/tests/pos/i12141.scala
+++ b/tests/pos/i12141.scala
@@ -1,0 +1,47 @@
+case class Test1(); case class Test2(); case class Test3();
+case class Test4(); case class Test5(); case class Test6();
+
+sealed abstract class DSL {
+  def cont [P1 >: this.type <: DSL, P2 <: DSL](continuation: => P2) =
+    Continue[P1, P2](() => this, () => continuation)
+}
+case class Continue [P1 <: DSL, P2 <: DSL](p1: () => P1, p2: () => P2) extends DSL
+
+trait More[-A] {}
+case class Out[C <: More[A], A](c: C, v: A) extends DSL
+case class Nop() extends DSL
+
+val decision1:Boolean = true;
+val decision2:Boolean = false;
+
+type P[
+ChanA <: More[Test1|Test2],
+ChanB <: More[Test3|Test4],
+ChanC <: More[Test5|Test6]] =
+ ((Out[ChanA,Test1] Continue ((Out[ChanB,Test3] Continue Nop)|(Out[ChanB,Test4] Continue Nop))) //works if remove first 'Continue Nop'
+ | (Out[ChanA,Test2] Continue ((Out[ChanC,Test5] Continue Nop)|(Out[ChanC,Test6] Continue Nop))))
+
+
+def p( chanA: More[Test1|Test2], chanB: More[Test3|Test4], chanC: More[Test5|Test6])
+ :P[chanA.type,chanB.type,chanC.type] ={
+      if(decision1){
+         Out(chanA,Test1()) cont {
+            if(decision2){
+               Out(chanB,Test3()) cont Nop() //works if replace with 'Out(chanB,Test3())'
+            }
+            else{
+               Out(chanB,Test4()) cont Nop()
+            }
+         }
+      }
+      else{
+         Out(chanA,Test2()) cont {
+            if(decision2){
+               Out(chanC,Test5()) cont Nop()
+            }
+            else{
+               Out(chanC,Test6()) cont Nop()
+            }
+         }
+      }
+   }

